virtualBridge.py

```diff
@@ -34,40 +36,63 @@ def __init__(self, h: host.Host, config: BridgeConfig):
+            ip = cfg.ip
+            mac = cfg.mac
+            name = cfg.name
+
+            host_xml = f"<host mac='{mac}' name='{name}' ip='{ip}'/>"
+            logger.info(f"Creating static DHCP entry for VM {name}, ip {ip} mac {mac}")
+            cmd = f"virsh net-update default add ip-dhcp-host \"{host_xml}\" --live --config"
             self.hostconn.run_or_die(cmd)
 
-        cmd = "virsh net-dhcp-leases default"
-        ret = self.hostconn.run(cmd)
-        # Look for "{name} " in the output. The space is intended to differentiate between "bm-worker-2 " and e.g. "bm-worker-20"
-        if f"{name} " in ret.out:
-            logger.error(f"Error: {name} found in dhcp leases")
-            logger.error("To fix this, run")
-            logger.error("\tvirsh net-destroy default")
-            logger.error("\tRemove wrong entries from /var/lib/libvirt/dnsmasq/virbr0.status")
-            logger.error("\tvirsh net-start default")
-            logger.error("\tsystemctl restart libvirt")
-            sys.exit(-1)
-
-        host_xml = f"<host mac='{mac}' name='{name}' ip='{ip}'/>"
-        logger.info(f"Creating static DHCP entry for VM {name}, ip {ip} mac {mac}")
-        cmd = f"virsh net-update default add ip-dhcp-host \"{host_xml}\" --live --config"
-        self.hostconn.run_or_die(cmd)
+    def remove_dhcp_entries(self, vms: list[NodeConfig]) -> None:
+        xml_str = self.hostconn.run("virsh net-dumpxml default").out
+        q = et.fromstring(xml_str)
+        removed_macs = []  # type: list[str]
+        names = [x.name for x in vms]
+        ips = [x.ip for x in vms]
+        for e in q[-1][0][1:]:
+            if e.attrib["name"] in names or e.attrib["ip"] in ips:
+                mac = e.attrib["mac"]
+                name = e.attrib["name"]
+                ip = e.attrib["ip"]
+                pre = "virsh net-update default delete ip-dhcp-host"
+                cmd = f"{pre} \"<host mac='{mac}' name='{name}' ip='{ip}'/>\" --live --config"
+                logger.info(self.hostconn.run(cmd))
+                removed_macs.append(mac)
+
+        fn = "/var/lib/libvirt/dnsmasq/virbr0.status"
+        with open(fn) as f:
+            contents = f.read()
+
+        if contents:
+            j = json.loads(contents)
+            names = [x.name for x in vms]
+            logger.info(f'Cleaning up {fn}')
+            logger.info(f'removing hosts with mac in {removed_macs} or name in {names}')
+            filtered = []
+            for entry in j:
+                if entry["mac-address"] in removed_macs:
+                    logger.info(f'Removed host with mac {entry["mac-address"]}')
+                    continue
+                if "hostname" in entry and entry["hostname"] in names:
+                    logger.info(f'Removed host with name {entry["hostname"]}')
+                    continue
+                logger.info(f'Kept entry {entry}')
+                filtered.append(entry)
+
```

Comment:
This is highly nested. Could we put this for loop inside its own function?

Reply:
Will do in updated PR.